### DISCUSSION TITLE
Allow vcpkg export with manifest-based projects

### DISCUSSION THREAD

**Discussion by morphogencc** [microsoft/vcpkg#24668] [Ideas]
****

**Is your feature request related to a problem? Please describe.**
I use manifests to track my dependencies and automatically build them in Visual Studio.  I've recently started doing Continuous Integration with GitHub Actions, and right now I build all dependencies using vcpkg in the Docker container:

```
        name: Add MSBuild to PATH
        uses: microsoft/setup-msbuild@v1.0.2

      - name: Install packages and dependencies
        run: |
          bootstrap-vcpkg
          vcpkg install --triplet "x64-windows-static" --x-manifest-root .
          vcpkg integrate install
      - name: Build project
        working-directory: ${{env.GITHUB_WORKSPACE}}
        run: msbuild /m /p:Configuration=${{env.BUILD_CONFIGURATION}} /p:Platform=${{env.BUILD_PLATFORM}} ${{env.SOLUTION_FILE_PATH}}
```

However, even though my projects only directly use 3-5 dependencies, once *those* libraries' dependencies are built, vcpkg is responsible for building 50-70 packages.  This takes upwards of 90 minutes to build in the docker containers, which is a lot longer than I'd prefer.

NuGet is pretty much built to solve this problem, so I looked into using it with vcpkg's export function to load the built binaries into an archive and load them to GitHub.  Unfortunately when I try it:

```
PS D:\Projects\myProjecct> .\dependencies\vcpkg\vcpkg.exe export --nuget
vcpkg export does not support manifest mode, in order to allow for future design considerations. You may use export in classic mode by running vcpkg outside of a manifest-based project.
Note: Updating vcpkg by rerunning bootstrap-vcpkg may resolve this failure.
```

**Proposed solution**
I'd like for `vcpkg export` to support exporting the binaries for dependencies built with the vcpkg manifest system, just like it works in classic mode. This way I could build a NuGet package, load it to my repository as a package, and then use it for automated testing of my C++ projects.

**Comment by autoantwort** [comment]
Maybe have a look at https://github.com/microsoft/vcpkg/blob/master/docs/users/binarycaching.md#github-packages

**Comment by morphogencc** [comment]
that looks really promising -- though it looks like that's just for restoring the package via NuGet and Github Packages, right?  I don't believe that example shows how the actually create the binary cache and push it to GitHub Packages.

**Comment by michaeltryby** [comment]
It would be great if I could specify a package store where my dependencies are cached in the manifest file. This seems obvious, so there must be a reason  why it hasn’t been implemented.

**Comment by morphogencc** [comment]
Agreed; the binary caching mode above would be super helpful it there was also a way to automatically *generate* the dependencies and store them.

**Reply by matthewclendening** [reply]
Isn't binary caching on by default? (vcpkg will cache on the local system). If you specified your own cache location(either through the  VCPKG_BINARY_SOURCES environment variable or the command line), would that be what you want?

**Comment by michaeltryby** [comment]
The use case I had in mind was an ephemeral system like a CI workflow. Although I do agree that the current binary cache mechanism works well when building locally.

**Reply by autoantwort** [reply]
You can also use the binary cache in a ci workflow. There are a lot of supported caching backends.

**Reply by michaeltryby** [reply]
I have configured binary cache on several CI systems. In my experience it leaves a lot to be desired. It is much easier and straight forward to fetch a package from a binary store.

**Comment by AugP** [comment]
Revisiting this to add that we have added vcpkg export support for manifest mode in the 2023.10.19 release. See our blog post for details: https://devblogs.microsoft.com/cppblog/vcpkg-2023-10-19-release-export-for-manifests-documentation-improvements-and-more/
